### PR TITLE
Reconciler extensions for Identity Center integration

### PR DESCRIPTION
**Pull request by tcsc**
This patch extends the existing Reconciler in 2 ways:
- Makes the resource key type generic, and
- Adds an optional custom comparison function

The Identity Center integration interacts with (and reconciles) several different
resource types, and using the builtin `string` for all of the ID values is just
asking for the wrong ID to get put in the wrong service. To mitigate this, both
the Identity Center integration and SCIM use strongly-types IDs in all of their
services, making the use of an incorrect ID a compile-time error.

Adding the generic key type allows the Identity Center integration and SCIM
clients use the Reconciler without a heap of casting or map rebuilding.

Prior to this patch the only way to specify a custom comparision routine is to
implement `compare.IsEqual[T]` on the target resource types. The IC integration
uses resources are defined outside the IC integration package, and the new RFD153-
style resources forbid adding custom methods to protobuf-generated types in their
packages.

The addition of a custom comparison function allows a Reconciler to specify a
custom comparison routine that will work with RDF153-style resources without
having to deal with wrapper types and avoiding the same sort of casting/map-rebuilding
issues as the genric resource keys.

This patch also supplies backwards-compatibilty wrappers, in order to avoid having
to update the many, many Reconciler useage sites.

### CODE REVIEW

File: lib/services/reconciler_test.go [hunk 1]
```diff
@@ -182,9 +229,9 @@ func TestReconciler(t *testing.T) {
 			// Reconcile and make sure we got all expected callback calls.
 			err = reconciler.Reconcile(context.Background())
 			require.NoError(t, err)
-			require.Equal(t, test.onCreateCalls, onCreateCalls)
-			require.Equal(t, test.onUpdateCalls, onUpdateCalls)
-			require.Equal(t, test.onDeleteCalls, onDeleteCalls)
+			require.ElementsMatch(t, test.onCreateCalls, onCreateCalls)
```
Comment (tcsc) [inline lib/services/reconciler_test.go:232]: Note for reviewers - I've relaxed the conditions on the call tests to allow the callback invocations to be in arbitrary order. I am reasonably sure that the order of callbacks is not defined or important, but will happily listen to anyone that says so.

File: lib/services/reconciler.go [hunk 2]
```diff
@@ -97,16 +146,25 @@ func NewReconciler[T any](cfg ReconcilerConfig[T]) (*Reconciler[T], error) {
...
+}
+
+// GenericReconciler reconciles currently registered resources with new
+// resources and creates/updates/deletes them appropriately.
+//
 // It's used in combination with watchers by agents (app, database, desktop)
 // to enable dynamically registered resources.
-type Reconciler[T any] struct {
-	cfg ReconcilerConfig[T]
+type GenericReconciler[K comparable, T any] struct {
```
Comment (greedy52) [inline lib/services/reconciler.go:160]: Add some test for `comparable K` instead of `string`?

File: lib/services/reconciler.go [hunk 3]
```diff
@@ -28,8 +28,15 @@ import (
...
 
+// Matcher is used by reconciler to match resources.
+type Matcher[T any] func(T) bool
+
 // ReconcilerConfig is the resource reconciler configuration.
 type ReconcilerConfig[T any] struct {
+	// NOTE: This struct *must* maintain the same layout as
+	//       GenericReconcilerConfig[string, T] in order for automatic
+	//       conversions to work
+
```
Comment (greedy52) [inline lib/services/reconciler.go:39]:
```suggestion
// ReconcilerConfig is the resource reconciler configuration.
type ReconcilerConfig[T any] GenericReconcilerConfig[string, T]
```
this should work?

**Review by greedy52 — COMMENTED** [review]
> This patch also supplies backwards-compatibilty wrappers, in order to avoid having
to update the many, many Reconciler useage sites.

Personally i would avoid having another reconciler type and then just update all the places. not sure what others thoughts though.

File: lib/services/reconciler.go (hunk 3 above)
Comment (tcsc) [reply lib/services/reconciler.go:39]: Ah cute. I though I'd tried it before, but perhaps I tried an alias instead.

**Comment by tcsc** [comment]
> Personally i would avoid having another reconciler type and then just update all the places. not sure what others thoughts though.

I count 19 uses to update, each of which has at least 2 places that need updating (struct declaration and construction) - that's a lot of noise to add to this PR.

That said, I'm happy to go back and update them once this patch is in. This way it won't break `e` in the meantime.

**Review by r0mant — APPROVED** [review]
Nice 👍

**Review by flyinghermit — APPROVED** [review]
lgtm

**Review by greedy52 — APPROVED** [review]

**Pull request merged**